analyzers/tests/SonarAnalyzer.UnitTest/TestCases/InfiniteRecursion.Roslyn.cs

```diff
@@ -0,0 +1,357 @@
+            {
+                return infiniteRecursionField?.Prop1;
+            }
+        }
+
+        void InternalRecursion(int i)
+        {
+        start:
+            goto end;
+        end:
+            goto start; // Compliant FN
+
+            switch (i)
+            {
+                case 1:
+                    goto default;
+                case 2:
+                    break;
+                default:
+                    goto case 1; // Compliant, already not reachable
+            }
+        }
+
+        int Pow2(int num, int exponent)
+        {
+            if (exponent > 1)
+            {
+                num = num * Pow2(num, exponent - 1);
+            }
+            return num;
+        }
+
+        void Generic<T>() // Noncompliant
+        {
+            Generic<T>();
+        }
+
+        void Generic2<T>() // Compliant
+        {
+            Generic2<int>();
+        }
+
+        int Power(int num, int exponent) // Noncompliant
+        {
+            try
+            {
+                num = num * Power(num, exponent - 1);
+                return num;  // this is never reached
+            }
+            catch (Exception)
+            {
+                throw;
+            }
+        }
+
+        int FixIt(int a)    // FN - Two methods calling each other are not recognized
+        {
+            return UpdateIt(a);
+        }
+
+        int UpdateIt(int a)
+        {
+            return FixIt(a);
+        }
+
+        int CSharp8_SwitchExpressions_OK(int a)
+        {
+            return a switch
+            {
+                0 => 1,
+                1 => 0,
+                _ => CSharp8_SwitchExpressions_OK(a) % 2
+            };
+        }
+
+        int CSharp8_SwitchExpressions_Bad(int a)    // Noncompliant
+        {
+            return a switch
+            {
+                0 => CSharp8_SwitchExpressions_Bad(a + 1),
+                1 => CSharp8_SwitchExpressions_Bad(a - 1),
+                _ => CSharp8_SwitchExpressions_Bad(a) % 2
+            };
+        }
+
+        int CSharp8_StaticLocalFunctions_OK(int a)
+        {
+            static int Calculate(int a, int b) => a + b + 1;
+
+            return Calculate(a, 1);
+        }
+
+        int CSharp8_StaticLocalFunctions_Bad(int a, int b)
+        {
+            static int Calculate(int a, int b) => Calculate(a, b) + 1;  //Noncompliant
+
+            return Calculate(a, b);
+        }
+
+        int CSharp8_StaticLocalFunctions_FN(int a, int b)
+        {
+            static int Add(int a, int b) => Fix(a, b);  // FN - Two methods calling each other are not recognized
+            static int Fix(int a, int b) => Add(a, b);
+
+            return Add(a, b);
+        }
+
+        void Foo()          // Compliant FN
+        {
+            Bar();
+            void Bar()      // Compliant FN
+            {
+                Baz();
+
+                void Baz()  // Noncompliant
```

Comment:
The case you've stated is valid, though the case I've supplied in test, tests the ability to retrieve and "not fail" on nested functions.

Reply:
Applied some renamings

Reply:
This still isn't clear why it is noncompliant. Is this detected, because `NestedLocalFunction()`:
* Invoking `MethodWithNestedLocalFunctions();` causes the recursion?
* Invoking `LocalFunction();` causes the recursion?
* Invoking `NestedLocalFunction();` causes the recursion?
* Or all of them?

And if we break implementation for one of them, this test case will still pass?

Reply:
This one wasn't addressed

Reply:
It is non-compliant since  NestedLocalFunction calls itself recursively.

Tried to make FN more understandable in a separate example.